plugins/module_utils/csv.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2021, Andrew Pantuso (@ajpantuso) <ajpantuso@gmail.com>
+# Copyright: (c) 2018, Dag Wieers (@dagwieers) <dag@wieers.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+import csv
+from io import BytesIO, StringIO
+
+from ansible.module_utils._text import to_native
+from ansible.module_utils.six import PY3
+
+
+class CustomDialectFailureError(Exception):
+    pass
+
+
+class DialectNotAvailableError(Exception):
+    pass
+
+
+CSVError = csv.Error
+
+
+def initialize_dialect(dialect, **kwargs):
+    # Add Unix dialect from Python 3
+    class unix_dialect(csv.Dialect):
+        """Describe the usual properties of Unix-generated CSV files."""
+        delimiter = ','
+        quotechar = '"'
+        doublequote = True
+        skipinitialspace = False
+        lineterminator = '\n'
+        quoting = csv.QUOTE_ALL
+
+    csv.register_dialect("unix", unix_dialect)
+
+    if dialect not in csv.list_dialects():
+        raise DialectNotAvailableError("Dialect '%s' is not supported by your version of python." % dialect)
+
+    # Create a dictionary from only set options
+    dialect_params = dict((k, v) for k, v in kwargs.items() if v is not None)
+    if dialect_params:
+        try:
+            csv.register_dialect('custom', dialect, **dialect_params)
+        except TypeError as e:
+            raise CustomDialectFailureError("Unable to create custom dialect: %s" % to_native(e))
+        dialect = 'custom'
+
+    return dialect
+
+
+def read_csv(data, dialect, fieldnames=None):
+
+    data = to_native(data, errors='surrogate_or_strict')
```

Comment:
In the original code, `data = to_text(data, errors='surrogate_or_strict')` was used in case of PY3 and nothing in case of PY2. What is the reason behind doing it differently here?

Reply:
The original module was leveraging the fact that it knew the file was being read in binary format before processing.
However to make this compatible with input passed to a filter (And other general use cases) the read_csv function needs to also convert text input in python 2 using [to_native](https://github.com/ansible/ansible/blob/089d0a0508a470799d099d95fc371e66756a00b3/lib/ansible/module_utils/common/text/converters.py#L256-L259) which directs to to_bytes in that case.
Moreover to_bytes just returns binary data untouched. See [here](https://github.com/ansible/ansible/blob/089d0a0508a470799d099d95fc371e66756a00b3/lib/ansible/module_utils/common/text/converters.py#L97)